server/src/main/java/org/opensearch/search/query/QueryCollectorContextSpecFactory.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.query;
+
+import org.apache.lucene.search.Query;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+
+/**
+ *  interface of QueryCollectorContext spec factory
+ */
+public interface QueryCollectorContextSpecFactory {
+    /**
+     * check if query is supported to initialize the factory
+     * @param query sent in the search request
+     * @return true if query satisfies the factory initialization criteria
+     */
+    boolean supports(Query query);
+
+    /**
+     *
+     * @param searchContext context needed to create collector context spec
+     * @param hasFilterCollector flag true if filter collector there
+     * @return QueryCollectorContextSpec
+     * @throws IOException
+     */
+    QueryCollectorContextSpec createQueryCollectorContextSpec(SearchContext searchContext, boolean hasFilterCollector) throws IOException;
```

Comment:
Let's return `Optional` and accept `QueryCollectorArguments` so to leave a way to add more arguments in the future without modifying the method signature:

```suggestion
    Optional<QueryCollectorContextSpec> createQueryCollectorContextSpec(SearchContext searchContext, QueryCollectorArguments arguments) throws IOException;
```

Reply:
I think we should not keep it Optional the reason being suppose there are multiple factories we only load the one which satisfies a specific query type or condition. Then in that case we definitely need a QueryCollectorContextSpec.

Reply:
The `searchContext.query()` provides access to the query being processed, if there is no custom query spec, the default is appropriate, right?

Reply:
@reta there is one condition when we have to throw error. 
1. if factory is created and no collectorContextSpec is created due to any exception or may be some error.

In that case, we cannot fall back on default implementation.

Reply:
> if factory is created and no collectorContextSpec is created due to any exception or may be some error.

@vibrantvarun it does not matter in this case - the query execution will fail and we don't need to proceed with default or alike